### PR TITLE
fix: compile production activation schedule

### PR DESCRIPTION
**Pull request by 2tbmz9y2xt-lang**
## Pull request overview

Makes production rotation activation derive only from a compiled per-network schedule artifact, so `mainnet`/`testnet` no longer bootstrap activation state from local `suite_registry` alone.

This PR also carries a small legacy-exposure contract hardening follow-up that landed while tightening the same review wave: the frozen hook-vector truth table is now asserted exactly in Go/Rust tests, and the Python contract checker now enforces the example fixture's cross-field semantics instead of only schema shape.

**Changes:**
- add canonical `production_rotation_schedule_v1.json` with explicit empty schedule entries for `mainnet` and `testnet`
- wire Go and Rust node startup to use the compiled schedule as the only production activation authority
- keep production local `rotation_descriptor` rejected and ignore local `suite_registry` entirely on production
- add matched Go/Rust loader and startup tests for empty schedule, malformed schedule, missing keys, unknown keys, non-finite `sunset_height`, and non-production caller misuse
- harden frozen legacy-exposure contract checks by pinning the exact canonical hook truth table in Go/Rust parity tests and by validating `legacy_exposure_total` / `watched_legacy_suite_ids` / `legacy_suite_reports` consistency in `tools/check_legacy_exposure_report_contract.py`
- update fixture changelog for the new protocol artifact

**Verification:**
- `go test ./node ./cmd/rubin-node`
- `cargo test --manifest-path clients/rust/Cargo.toml -p rubin-node`
- `python3 tools/check_legacy_exposure_report_contract.py`
- `gofmt -w clients/go/node/config.go clients/go/node/production_rotation_schedule_test.go clients/go/cmd/rubin-node/main_test.go`
- `cargo fmt --manifest-path clients/rust/Cargo.toml --all`
- `git diff --check`

Refs: #1114

### CODE REVIEW

File: clients/go/node/production_rotation_schedule.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+	"github.com/2tbmz9y2xt-lang/rubin-protocol/clients/go/consensus"
+)
+
+const productionRotationScheduleVersion = 1
+const productionRotationScheduleErrStem = "production_rotation_schedule"
+
+// Derived runtime copy of conformance/fixtures/protocol/production_rotation_schedule_v1.json.
+// Go embed cannot read a parent-path artifact directly, so tests keep this byte-equivalent
+// to the canonical protocol fixture.
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule.go]: Fixed in `3296657`: the comment now says `JSON-equivalent` rather than `byte-equivalent`, matching the actual `json.Compact`-based drift check.

File: clients/go/node/production_rotation_schedule_test.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	_, _, err := loadCompiledProductionRotationScheduleFromJSONWithRegistry([]byte(`{
+		"version": 1,
+		"networks": {"mainnet": [], "testnet": null}
+	}`), consensus.DefaultSuiteRegistry())
+	if err == nil {
+		t.Fatal("expected malformed descriptor rejection")
+	}
+	if got, want := err.Error(), `production_rotation_schedule: networks.mainnet: json: cannot unmarshal array into Go value of type node.RotationConfigJSON`; got != want {
+		t.Fatalf("error=%q, want %q", got, want)
+	}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule_test.go:175]: Fixed in `3296657`: the test no longer asserts the unstable raw `encoding/json` error string. It now checks the stable schedule/path stem and uses `errors.As(..., *json.UnmarshalTypeError)` plus `Value == "array"`.

File: conformance/fixtures/CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,21 @@ Policy:
...
+- нужен один compiled authoritative per-network schedule artifact с явным empty-schedule поведением для `mainnet/testnet`;
+- локальный `rotation_descriptor` остаётся non-production only, а production activation authority теперь идёт из общего protocol fixture.
+
+Инструменты:
+- новый protocol fixture `conformance/fixtures/protocol/production_rotation_schedule_v1.json`,
+- Go/Rust node loader tests,
+- проверка через `go test ./node ./cmd/rubin-node` и `cargo test -p rubin-node`.
+
+Изменённые fixtures:
+- `protocol/production_rotation_schedule_v1.json` (new)
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply conformance/fixtures/CHANGELOG.md:27]: Fixed in `3296657`: `conformance/fixtures/CHANGELOG.md` now explicitly covers shared protocol artifacts under `conformance/fixtures/protocol/*.json`, so this entry matches the documented changelog scope.

File: clients/go/node/production_rotation_schedule.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+func parseProductionRotationScheduleDescriptor(
+	raw json.RawMessage,
+	network string,
+	registry *consensus.SuiteRegistry,
+) (*consensus.CryptoRotationDescriptor, error) {
+	if bytes.Equal(bytes.TrimSpace(raw), []byte("null")) {
+		return nil, nil
+	}
+	var descriptorJSON RotationConfigJSON
+	if err := decodeSingleJSONValue(raw, &descriptorJSON); err != nil {
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule.go]: Fixed in `3296657`: Go no longer decodes schedule descriptors directly into `RotationConfigJSON`. It now parses through a pointer-based wire struct, rejects missing required fields before descriptor validation, and has a regression test for missing `create_height`, matching the Rust fail-closed behavior.

File: clients/go/node/production_rotation_schedule_test.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+func TestEmbeddedProductionRotationScheduleMatchesCanonicalFixture(t *testing.T) {
+	raw, err := os.ReadFile(productionRotationScheduleRepoPath(
+		"conformance",
+		"fixtures",
+		"protocol",
+		"production_rotation_schedule_v1.json",
+	))
+	if err != nil {
+		t.Fatalf("ReadFile(canonical fixture): %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule_test.go:70]: Fixed in `30d1ccf`: the test now derives the canonical fixture path from `runtime.Caller` and validates it with `os.Stat` before `ReadFile`, so it no longer depends on the test binary working directory or a relative path layout. I did not switch this check to `t.Skip`, because this fixture is the canonical production schedule artifact for the repo and a missing file here is real fixture/layout drift that should fail closed rather than silently skip.

File: clients/rust/crates/rubin-node/src/production_rotation_schedule.rs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+#[derive(Clone, Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+struct ProductionRotationDescriptorWire {
+    name: String,
+    old_suite_id: u8,
+    new_suite_id: u8,
+    create_height: u64,
+    spend_height: u64,
+    #[serde(default)]
+    sunset_height: u64,
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/rust/crates/rubin-node/src/production_rotation_schedule.rs]: Fixed in `2792da6`: Rust now parses `sunset_height: null` via `Option<u64>`, matching Gos parser surface instead of failing deserialization early. Both clients still reject the resulting production descriptor with the same finite-H4 validation error, and the new regression test pins that exact behavior so this no longer creates a cross-client divergence.

File: clients/go/node/production_rotation_schedule.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+) *consensus.SuiteRegistry {
+	paramsByID := map[uint8]consensus.SuiteParams{
+		consensus.SUITE_ID_ML_DSA_87: defaultSuiteRegistryParams(),
+	}
+	for _, descriptor := range descriptors {
+		if descriptor == nil {
+			continue
+		}
+		paramsByID[descriptor.OldSuiteID] = canonicalProductionScheduleSuiteParams(descriptor.OldSuiteID)
+		paramsByID[descriptor.NewSuiteID] = canonicalProductionScheduleSuiteParams(descriptor.NewSuiteID)
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule.go]: Fixed in `0d83062`: both clients now reject `old_suite_id/new_suite_id == 0x00` at the production schedule parse boundary, before derived registry construction. That keeps `SUITE_ID_SENTINEL` from self-registering and adds direct Go/Rust regression coverage for both `old_suite_id` and `new_suite_id`.

File: clients/rust/crates/rubin-node/src/production_rotation_schedule.rs [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+) -> SuiteRegistry {
+    let mut suites = BTreeMap::new();
+    suites.insert(
+        SUITE_ID_ML_DSA_87,
+        canonical_production_schedule_suite_params(SUITE_ID_ML_DSA_87),
+    );
+    for descriptor in [mainnet.as_ref(), testnet.as_ref()].into_iter().flatten() {
+        suites.insert(
+            descriptor.old_suite_id,
+            canonical_production_schedule_suite_params(descriptor.old_suite_id),
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/rust/crates/rubin-node/src/production_rotation_schedule.rs]: Fixed in `0d83062`: both clients now reject `old_suite_id/new_suite_id == 0x00` at the production schedule parse boundary, before derived registry construction. That keeps `SUITE_ID_SENTINEL` from self-registering and adds direct Go/Rust regression coverage for both `old_suite_id` and `new_suite_id`.

File: clients/rust/crates/rubin-node/src/production_rotation_schedule.rs [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+        );
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/rust/crates/rubin-node/src/production_rotation_schedule.rs]: Fixed in `0d83062`: both clients now reject `old_suite_id/new_suite_id == 0x00` at the production schedule parse boundary, before derived registry construction. That keeps `SUITE_ID_SENTINEL` from self-registering and adds direct Go/Rust regression coverage for both `old_suite_id` and `new_suite_id`.

File: clients/go/node/production_rotation_schedule.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+) *consensus.SuiteRegistry {
+	paramsByID := map[uint8]consensus.SuiteParams{
+		consensus.SUITE_ID_ML_DSA_87: canonicalProductionScheduleSuiteParams(consensus.SUITE_ID_ML_DSA_87),
+	}
+	for _, descriptor := range descriptors {
+		if descriptor == nil {
+			continue
+		}
+		paramsByID[descriptor.OldSuiteID] = canonicalProductionScheduleSuiteParams(descriptor.OldSuiteID)
+		paramsByID[descriptor.NewSuiteID] = canonicalProductionScheduleSuiteParams(descriptor.NewSuiteID)
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule.go]: Fixed in `0d83062`: both clients now reject `old_suite_id/new_suite_id == 0x00` at the production schedule parse boundary, before derived registry construction. That keeps `SUITE_ID_SENTINEL` from self-registering and adds direct Go/Rust regression coverage for both `old_suite_id` and `new_suite_id`.

**Comment by 2tbmz9y2xt-lang** [comment]
### Code review

**Verdict:** No high-confidence blocking issues for the stated scope (compiled production rotation schedule as sole production activation authority; Go/Rust + genesis wiring; parity tests).

**Reviewed (head `0d8306297830853d6564811dc81c94ba6da4527f`):**
- Production path rejects local `rotation_descriptor`; empty schedule slots yield nil rotation/registry so local `suite_registry` stays validated but non-authoritative ([config.go](https://github.com/2tbmz9y2xt-lang/rubin-protocol/blob/0d8306297830853d6564811dc81c94ba6da4527f/clients/go/node/config.go#L223-L264)).
- Strict JSON: outer + inner decode use `DisallowUnknownFields` in Go ([production_rotation_schedule.go](https://github.com/2tbmz9y2xt-lang/rubin-protocol/blob/0d8306297830853d6564811dc81c94ba6da4527f/clients/go/node/production_rotation_schedule.go#L46-L62)); Rust wire types use `deny_unknown_fields` ([production_rotation_schedule.rs](https://github.com/2tbmz9y2xt-lang/rubin-protocol/blob/0d8306297830853d6564811dc81c94ba6da4527f/clients/rust/crates/rubin-node/src/production_rotation_schedule.rs#L30-L44)).
- Empty-slot callers avoid returning the shared derived registry to prevent cross-network suite leakage; matches Rust `Option` mapping ([production_rotation_schedule.go](https://github.com/2tbmz9y2xt-lang/rubin-protocol/blob/0d8306297830853d6564811dc81c94ba6da4527f/clients/go/node/production_rotation_schedule.go#L268-L283), [production_rotation_schedule.rs](https://github.com/2tbmz9y2xt-lang/rubin-protocol/blob/0d8306297830853d6564811dc81c94ba6da4527f/clients/rust/crates/rubin-node/src/production_rotation_schedule.rs#L191-L206)).
- Drift control: Go embed vs canonical fixture equality test; Rust `include_str!` on the same fixture path.

**Note:** `ValidateConfig` always exercises schedule load on mainnet/testnet — acceptable; only a micro-latency concern if ever hot-path.

Workspace review: root `CLAUDE.md` / `AGENTS.md` — no PR-specific violations identified for this change set.

File: clients/go/node/production_rotation_schedule.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+			)
+		}
+	}
+	schedule := &productionRotationSchedule{
+		Version:  wire.Version,
+		Networks: make(map[string]*consensus.CryptoRotationDescriptor, len(wire.Networks)),
+	}
+	parsedDescriptors := make(map[string]*RotationConfigJSON, len(wire.Networks))
+	for _, network := range []string{"mainnet", "testnet"} {
+		raw, ok := wire.Networks[network]
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule.go]: Fixed in `d63ba81`: renamed the inner `raw` binding to `entryRaw` in `loadCompiledProductionRotationScheduleFromJSONWithRegistry`, removing the shadowing without changing schedule semantics or parser behavior.

File: clients/go/node/production_rotation_schedule.go [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+			network,
+		)
+	}
+	descriptor := schedule.Networks[network]
+	if descriptor == nil {
+		// A null slot means this network has no authoritative production activation state.
+		// Pre-rotation production callers still need the canonical default registry,
+		// but must not inherit foreign-network suites from the compiled schedule.
+		return nil, consensus.DefaultSuiteRegistry(), nil
+	}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/production_rotation_schedule.go:251]: Fixed in a41fd0c: production runtime parity now aligns at the node boundary. Go and Rust both use an explicit default pre-rotation runtime for empty production schedule slots, and both ignore local suite_registry on mainnet/testnet. Rust production_rotation_descriptor_for_network() still models the raw schedule slot as None, but build_suite_context_from_descriptor_with_production_lookup() now turns that into an explicit default SuiteContext, which is runtime-equivalent to the Go default rotation plus registry path.

File: clients/go/node/config.go [hunk 13]
```diff
@@ -343,6 +367,20 @@ func ValidateConfig(cfg Config) error {
...
+	}
+	if network == "mainnet" || network == "testnet" {
+		if cfg.RotationDescriptor != nil {
+			return errors.New(productionLocalRotationDescriptorErr)
+		}
+		// Production validation checks only the compiled schedule artifact.
+		if _, _, err := productionRotationDescriptorForNetwork(network); err != nil {
+			return err
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/config.go]: Updated the PR description to match the implemented semantics: on `mainnet`/`testnet`, local `suite_registry` is ignored entirely and never treated as production authority. The code in `ValidateConfig`/startup already matched that behavior; the stale text was the drift.

File: clients/rust/crates/rubin-node/src/genesis.rs [hunk 14]
```diff
@@ -290,14 +311,28 @@ fn build_suite_context_from_descriptor(
...
+            })),
+            // Production empty slots use an explicit default pre-rotation
+            // context and never consult local suite_registry input.
+            None => Ok(Some(crate::sync::SuiteContext {
+                rotation: Arc::new(DefaultRotationProvider),
+                registry: Arc::new(SuiteRegistry::default_registry()),
+            })),
+        };
+    }
+    let explicit_registry = build_suite_registry_from_json(suite_registry)?;
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/rust/crates/rubin-node/src/genesis.rs]: Updated the PR description to match the implemented semantics: on production networks, local `suite_registry` is ignored entirely rather than validated-and-discarded. That now matches both the Go and Rust production branches, which return before local `suite_registry` parsing.

File: clients/rust/crates/rubin-node/src/genesis.rs [hunk 15]
```diff
@@ -740,14 +804,244 @@ mod tests {
...
+                    .expect("time")
+                    .as_nanos()
+            ));
+            std::fs::create_dir_all(&dir).expect("mkdir");
+            let path = dir.join("genesis.json");
+            std::fs::write(
+                &path,
+                "{\
+                  \"chain_id_hex\":\"0x88f8a9acdeeb902e27aa2fdcb8c46ecf818bf68dec5273ec1bcc5084e2333103\",\
+                  \"suite_registry\":[{\"suite_id\":77,\"pubkey_len\":2592,\"sig_len\":4627,\"verify_cost\":0,\"alg_name\":\"ML-DSA-87\"}]\
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/rust/crates/rubin-node/src/genesis.rs]: Fixed in `f7d7bd2`: the Rust test now builds the malformed production `suite_registry` entry through `suite_registry_entry_json(77, ML_DSA_87_PUBKEY_BYTES, ML_DSA_87_SIG_BYTES, 0, "ML-DSA-87")` instead of hardcoded numeric literals, matching the shared-helper pattern used elsewhere.

File: clients/go/node/config.go [hunk 16]
```diff
@@ -343,6 +367,29 @@ func ValidateConfig(cfg Config) error {
 			return fmt.Errorf("mine_address must be 32 (key_id) or 33 (suite_id||key_id) bytes, got %d", len(raw))
 		}
 	}
+	network, err := canonicalConfigNetworkName(cfg.Network)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/config.go]: Fixed in `94716f1`: `ValidateConfig` now canonicalizes `cfg.Network` once at the top, reuses the normalized `network` value for the production branch, and no longer duplicates the same normalization call later in the function.

[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply clients/go/node/config.go]: Fixed in `94716f1`: `ValidateConfig` now captures the canonicalized network once and reuses that value for the production branch. The second `canonicalConfigNetworkName(cfg.Network)` call is gone, so the data flow is single-pass and there is no duplicate normalization left in this function.
